lib/AST/SubstitutionMap.cpp

```diff
@@ -70,10 +70,19 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
       genericEnv->mapTypeOutOfContext(archetype)->getCanonicalType());
   }
 
-  auto known = subMap.find(cast<GenericTypeParamType>(type));
+  auto genericParam = cast<GenericTypeParamType>(type);
+  auto known = subMap.find(genericParam);
   if (known != subMap.end() && known->second)
     return known->second;
 
+  // The generic parameter may have been made concrete by the generic signature,
```

Comment:
We have to do it lazily because the concrete types can end up referring to other generic type parameters. Which also means I need to deal with the recursion in ill-formed programs (hence the failures).